homeassistant/components/minecraft_server/api.py

```diff
@@ -0,0 +1,130 @@
+"""API for the Minecraft Server integration."""
+
+
+from dataclasses import dataclass
+from enum import StrEnum
+
+from mcstatus import BedrockServer, JavaServer
+from mcstatus.status_response import BedrockStatusResponse, JavaStatusResponse
+
+
+@dataclass
+class MinecraftServerData:
+    """Representation of Minecraft Server data."""
+
+    # Common data
+    latency: float
+    motd: str
+    players_max: int
+    players_online: int
+    protocol_version: int
+    version: str
+
+    # Data available only in 'Java Edition'
+    players_list: list[str] | None
+
+    # Data available only in 'Bedrock Edition'
+    edition: str | None
+    game_mode: str | None
+    map_name: str | None
+
+
+class MinecraftServerType(StrEnum):
+    """Enumeration of Minecraft Server types."""
+
+    JAVA_EDITION = "Java Edition"
+    BEDROCK_EDITION = "Bedrock Edition"
+
+
+class MinecraftServerAddressError(Exception):
+    """Raised when the input address is invalid."""
+
+
+class MinecraftServerConnectionError(Exception):
+    """Raised when no data can be fechted from the server."""
+
+
+class MinecraftServer:
```

Comment:
I like the idea! I don't really like the idea of splitting up the retrieving logic into the coordinator and this api class.

What if in the `__init__.py` you (use a helper to) try to lookup the server type, after which we pass the MCServer to the coordinator. With as added benefit, if the server can't be looked up, we can say that the config entry is not ready.

Then you can put the rest of the mapping logic in helpers to not make the coordinator one big pile of code.

Reply:
I think it would be nice if the coordinator shouldn't have to know what kind of server he is talking with and put those specifics away in functions, making the coordinator clean

Reply:
Good point, thanks :+1: 

I create now the API instance (`MinecraftServer`) in `__init__.py` and pass it to the coordinator. If a problem with the config entry is detected, `ConfigEntryError` is now raised (before, `HomeAssistantError` was raised instead in the coordinator).

I also reached now a state, where I cannot further simplify the coordinator. It's now as simple as it can get :smile: 

I still stick to the server type stored in the config entry data and not to detect the server type in `__init__.py` due to the possibly long waiting times I mentioned before.

BTW: `ConfigEntryNotReady` is already handled by `coordinator.async_config_entry_first_refresh`, called in `async_setup_entry.`